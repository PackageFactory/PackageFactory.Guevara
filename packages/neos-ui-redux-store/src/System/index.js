import {createAction} from 'redux-actions';

import {handleActions} from '@neos-project/utils-redux';

const BOOT = '@neos/neos-ui/System/BOOT';
const INIT = '@neos/neos-ui/System/INIT';
const READY = '@neos/neos-ui/System/READY';

//
// Export the action types
//
export const actionTypes = {
    BOOT,
    INIT,
    READY
};

const boot = createAction(BOOT);
const init = createAction(INIT, state => state);
const ready = createAction(READY);

//
// Export the actions
//
export const actions = {
    boot,
    init,
    ready
};

//
// Export the reducer
//
export const reducer = handleActions({
});

//
// Export the selectors
//
export const selectors = {};
