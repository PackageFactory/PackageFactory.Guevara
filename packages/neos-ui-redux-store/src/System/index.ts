import produce from 'immer';
import {$get} from 'plow-js';
import {action as createAction, ActionType as ActionTypeHelper} from 'typesafe-actions';

const BOOT = '@neos/neos-ui/System/BOOT';
const INIT = '@neos/neos-ui/System/INIT';
const READY = '@neos/neos-ui/System/READY';
const AUTHENTICATION_TIMEOUT = '@neos/neos-ui/System/AUTHENTICATION_TIMEOUT';
const REAUTHENTICATION_SUCCEEDED = '@neos/neos-ui/System/REAUTHENTICATION_SUCCEEDED';

//
// Export the action types
//
export const actionTypes = {
    AUTHENTICATION_TIMEOUT,
    BOOT,
    INIT,
    READY,
    REAUTHENTICATION_SUCCEEDED
};

interface State {
    readonly authenticationTimeout: boolean;
}

const boot = () => createAction(BOOT);
const init = (state: State) => createAction(INIT, state);
const ready = () => createAction(READY);
const authenticationTimeout = () => createAction(AUTHENTICATION_TIMEOUT);
const reauthenticationSucceeded = () => createAction(REAUTHENTICATION_SUCCEEDED);

//
// Export the actions
//
export const actions = {
    authenticationTimeout,
    boot,
    init,
    ready,
    reauthenticationSucceeded
};

export type ActionType = ActionTypeHelper<typeof actions>;

const defaultState: State = {
    authenticationTimeout: false
};

//
// Export the reducer
//
export const reducer = (state: State = defaultState, action: ActionType) => {
    return produce(state, draft => {
        switch (action.type) {
            case INIT:
                draft.authenticationTimeout = false;
                break;
            case AUTHENTICATION_TIMEOUT:
                draft.authenticationTimeout = true;
                break;
            case REAUTHENTICATION_SUCCEEDED:
                draft.authenticationTimeout = false;
                break;
        }
    });
};

//
// Export the selectors
//
export const selectors = {
    authenticationTimeout: (state: any) => $get(['system'], state).authenticationTimeout
};
