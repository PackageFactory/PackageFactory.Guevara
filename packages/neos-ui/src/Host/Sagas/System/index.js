import {take, put} from 'redux-saga/effects';
import {delay, discover} from 'Shared/Utilities/Promises';
import {actionTypes, actions} from 'Host/Redux/index';
import initializeJSAPI from 'API/index';

let injectStore = null;
export const getStore = discover(function * () {
    const store = yield new Promise(resolve => {
        injectStore = resolve;
    });

    return store;
});

export const getAppContainer = discover(function * () {
    const appContainer = yield new Promise(resolve => {
        document.addEventListener('DOMContentLoaded', () => {
            resolve(document.getElementById('appContainer'));
        });
    });

    return appContainer;
});

export const getCsrfToken = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.csrfToken;
});

export const getSystemEnv = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.env;
});

export const getServerState = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="initialState"]').innerHTML);
});

export const getConfiguration = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="configuration"]').innerHTML);
});

export const getMenu = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="menu"]').innerHTML);
});

export const getTranslations = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="translations"]').innerHTML);
});

export const getNeos = discover(function * () {
    const csrfToken = yield getCsrfToken;
    const systemEnv = yield getSystemEnv;

    const neos = initializeJSAPI(window, {
        csrfToken,
        systemEnv
    });

    return neos;
});

function * bootSaga(store) {
    injectStore(store);

    yield take(actionTypes.System.BOOT);

    const serverState = yield getServerState;

    //
    // Kick off initialization for redux, other sagas can
    // listen to this action and start their initialization tasks
    //
    yield put(actions.System.init(serverState));

    //
    // Just make sure that everybody does their initialization
    // homework
    //
    yield delay(0);

    //
    // Inform everybody, that we're ready to rock
    //
    yield put(actions.System.ready());
}

export const sagas = [
    bootSaga
];
