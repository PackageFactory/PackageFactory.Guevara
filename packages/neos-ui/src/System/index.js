import {discover} from '@neos-project/utils-helpers';
import {initializeJsAPI} from '@neos-project/neos-ui-backend-connector';

export const getAppContainer = discover(function * () {
    const appContainer = yield new Promise(resolve => {
        document.addEventListener('DOMContentLoaded', () => {
            resolve(document.getElementById('appContainer'));
        });
    });

    return appContainer;
});

export const getCsrfToken = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.csrfToken;
});

export const getSystemEnv = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.env;
});

export const getServerState = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="initialState"]').innerHTML);
});

export const getConfiguration = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="configuration"]').innerHTML);
});

export const getNodeTypes = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="nodeTypes"]').innerHTML);
});

export const getMenu = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="menu"]').innerHTML);
});

export const getTranslations = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="translations"]').innerHTML);
});

export const getNeos = discover(function * () {
    const csrfToken = yield getCsrfToken;
    const systemEnv = yield getSystemEnv;

    const neos = initializeJsAPI(window, {
        csrfToken,
        systemEnv
    });

    return neos;
});
