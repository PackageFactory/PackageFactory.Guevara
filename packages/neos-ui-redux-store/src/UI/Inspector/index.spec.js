import test from 'ava';
import Immutable, {Map} from 'immutable';

import {actionTypes, actions, reducer, selectors} from './index.js';

import {actionTypes as system} from '../../System/index';

test(`should export actionTypes`, t => {
    t.not(actionTypes, undefined);
    t.is(typeof (actionTypes.COMMIT), 'string');
    t.is(typeof (actionTypes.CLEAR), 'string');
    t.is(typeof (actionTypes.APPLY), 'string');
    t.is(typeof (actionTypes.DISCARD), 'string');
});

test(`should export action creators`, t => {
    t.not(actions, undefined);
    t.is(typeof (actions.commit), 'function');
    t.is(typeof (actions.clear), 'function');
    t.is(typeof (actions.apply), 'function');
    t.is(typeof (actions.discard), 'function');
});

test(`should export a reducer`, t => {
    t.not(reducer, undefined);
    t.is(typeof (reducer), 'function');
});

test(`should export selectors`, t => {
    t.not(selectors, undefined);
    t.is(typeof (selectors.transientValues), 'function');
    t.is(typeof (selectors.viewConfiguration), 'function');
});

test(`The reducer should return an Immutable.Map as the initial state.`, t => {
    const state = new Map({});
    const nextState = reducer(state, {
        type: system.INIT
    });

    t.true(nextState.get('ui').get('inspector') instanceof Map);
    t.true(nextState.get('ui').get('inspector').get('valuesByNodePath') instanceof Map);
});

test(`The "commit" action should store the last modification on the currently focused node.`, t => {
    const state = Immutable.fromJS({
        cr: {
            nodes: {
                focused: {
                    contextPath: '/my/path@user-foo'
                }
            }
        },
        ui: {
            inspector: {
                valuesByNodePath: {}
            }
        }
    });
    const nextState1 = reducer(state, actions.commit('test', 'value'));
    const nextState2 = reducer(state, actions.commit('test', 'another value'));
    const nextState3 = reducer(nextState1, actions.commit('test', 'another value'));
    const nextState4 = reducer(nextState1, actions.commit('test', 'another value', {some: 'hook'}));

    t.deepEqual(nextState1.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/path@user-foo': {
            test: {
                value: 'value'
            }
        }
    });
    t.deepEqual(nextState2.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/path@user-foo': {
            test: {
                value: 'another value'
            }
        }
    });
    t.deepEqual(nextState3.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/path@user-foo': {
            test: {
                value: 'another value'
            }
        }
    });
    t.deepEqual(nextState4.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/path@user-foo': {
            test: {
                value: 'another value',
                hooks: {
                    some: 'hook'
                }
            }
        }
    });
});

test(`The "clear" action should remove pending changes for the currently focused node.`, t => {
    const state = Immutable.fromJS({
        cr: {
            nodes: {
                focused: {
                    contextPath: '/my/path@user-foo'
                }
            }
        },
        ui: {
            inspector: {
                valuesByNodePath: {
                    '/my/path@user-foo': {
                        test1: {
                            value: 'value1'
                        },
                        test2: {
                            value: 'value2'
                        },
                        test3: {
                            value: 'value3'
                        }
                    },
                    '/my/other/path@user-foo': {
                        test4: {
                            value: 'value4'
                        }
                    }
                }
            }
        }
    });
    const nextState1 = reducer(state, actions.clear());
    const nextState2 = reducer(nextState1, actions.clear());

    t.deepEqual(nextState1.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/other/path@user-foo': {
            test4: {
                value: 'value4'
            }
        }
    });
    t.deepEqual(nextState2.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/other/path@user-foo': {
            test4: {
                value: 'value4'
            }
        }
    });
});

test(`The "discard" action should remove pending changes for the currently focused node.`, t => {
    const state = Immutable.fromJS({
        cr: {
            nodes: {
                focused: {
                    contextPath: '/my/path@user-foo'
                }
            }
        },
        ui: {
            inspector: {
                valuesByNodePath: {
                    '/my/path@user-foo': {
                        test1: {
                            value: 'value1'
                        },
                        test2: {
                            value: 'value2'
                        },
                        test3: {
                            value: 'value3'
                        }
                    },
                    '/my/other/path@user-foo': {
                        test4: {
                            value: 'value4'
                        }
                    }
                }
            }
        }
    });
    const nextState1 = reducer(state, actions.discard());
    const nextState2 = reducer(nextState1, actions.discard());

    t.deepEqual(nextState1.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/other/path@user-foo': {
            test4: {
                value: 'value4'
            }
        }
    });
    t.deepEqual(nextState2.get('ui').get('inspector').get('valuesByNodePath').toJS(), {
        '/my/other/path@user-foo': {
            test4: {
                value: 'value4'
            }
        }
    });
});
