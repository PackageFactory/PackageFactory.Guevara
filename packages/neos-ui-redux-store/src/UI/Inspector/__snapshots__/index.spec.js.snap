// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The "clear" action should remove pending changes for the currently focused node. 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/other/path@user-foo": Immutable.Map {
          "test4": Immutable.Map {
            "value": "value4",
          },
        },
      },
      "shouldPromptToHandleUnappliedChanges": false,
    },
  },
}
`;

exports[`The "clear" action should remove pending changes for the currently focused node. 2`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/other/path@user-foo": Immutable.Map {
          "test4": Immutable.Map {
            "value": "value4",
          },
        },
      },
      "shouldPromptToHandleUnappliedChanges": false,
    },
  },
}
`;

exports[`The "commit" action should ignore the last modification, if it doesn't differ from the current node state 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
          "properties": Immutable.Map {
            "title": "Foo",
          },
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {},
    },
  },
}
`;

exports[`The "commit" action should store the last modification on the currently focused node. 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "test": Immutable.OrderedMap {
            "value": "value",
          },
        },
      },
    },
  },
}
`;

exports[`The "commit" action should store the last modification on the currently focused node. 2`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "test": Immutable.OrderedMap {
            "value": "another value",
          },
        },
      },
    },
  },
}
`;

exports[`The "commit" action should store the last modification on the currently focused node. 3`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "test": Immutable.OrderedMap {
            "value": "another value",
          },
        },
      },
    },
  },
}
`;

exports[`The "commit" action should store the last modification on the currently focused node. 4`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "test": Immutable.OrderedMap {
            "value": "another value",
            "hooks": Immutable.OrderedMap {
              "some": "hook",
            },
          },
        },
      },
    },
  },
}
`;

exports[`The "discard" action should remove pending changes for the currently focused node. 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/other/path@user-foo": Immutable.Map {
          "test4": Immutable.Map {
            "value": "value4",
          },
        },
      },
      "shouldPromptToHandleUnappliedChanges": false,
    },
  },
}
`;

exports[`The "discard" action should remove pending changes for the currently focused node. 2`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "nodes": Immutable.Map {
      "byContextPath": Immutable.Map {
        "/my/path@user-foo": Immutable.Map {
          "contextPath": "/my/path@user-foo",
        },
      },
      "focused": Immutable.Map {
        "contextPath": "/my/path@user-foo",
      },
    },
  },
  "ui": Immutable.Map {
    "inspector": Immutable.Map {
      "valuesByNodePath": Immutable.Map {
        "/my/other/path@user-foo": Immutable.Map {
          "test4": Immutable.Map {
            "value": "value4",
          },
        },
      },
      "shouldPromptToHandleUnappliedChanges": false,
    },
  },
}
`;
